### ISSUE TITLE
Best way to write tests for EPMs?

### ISSUE THREAD

**Issue by crinklywrappr** [elves/elvish#1427]
My [little epm](https://github.com/crinklywrappr/rivendell) is outdated and I would like to remedy that.  I also want to write tests to prevent regressions and help document proper use in the README.

There is a little test-runner built into the epm, and I noticed that I wasn't the only person to do this.  Here is another: https://github.com/zzamboni/elvish-modules/blob/master/test.org

This spurs a few questions.

1. does elvish include a test runner for epm's? 
2.  should it?
3. can elvish resolve dependencies in epm's?  
4. should it?

AFAIK the answers to 1 & 2 are _no_.  I'm happy to be wrong,

**Comment by zzamboni** [comment]
The answer to (1) is **no**. The answer to (2) I guess is **it depends**, it might not be a bad idea to include some tests to make sure things don't break across updates. Main thing would be to define a convention so that the `epm` module can trigger the tests when a module is installed or updated (might also help identify missing external binaries used by the module, for example). @xiaq any thoughts?

The answers to (3) and (4) are **yes**, dependencies can be specified in the `metadata.json` file, see https://elv.sh/ref/epm.html#epmmetadata

**Comment by krader1961** [comment]
I'm slightly surprised this appears to be the first time this has been asked. It certainly would be helpful if there were a framework analogous to `go test` for writing and running Elvish unit tests. Regardless of whether or not the Elvish code is published and installable via `epm`. In the short term the primary purpose of such a mechanism would be to make it easy to verify if an Elvish module works correctly with a specific Elvish version given that Elvish is undergoing rapid evolution with many backward incompatible changes. Someone using `epm` to install a module may be using an Elvish version newer than what the author of the module was using. Leading to compatibility problems. Long term it is clear that making it easy to write and execute Elvish unit tests is a "best practice".

**Comment by zzamboni** [comment]
I agree it would be great to have a good test framework for Elvish code. I wrote a simple one some time ago: https://github.com/zzamboni/elvish-modules/blob/master/test.org, but even I haven't used it that much. The best example of its use is my `util` module, where most functions have tests: https://github.com/zzamboni/elvish-modules/blob/master/util_test.elv

Mostly I think it would be a matter of agreeing on a simple, loosely-coupled convention for test execution. For example: after installing or upgrading a module, `epm` could automatically run any `*-test.elv` files found in the package.

The next question is: what happens if the test fails? Let the user figure it out? Roll back the install? (If so, how?) Something else?

**Comment by crinklywrappr** [comment]
> Mostly I think it would be a matter of agreeing on a simple, loosely-coupled convention for test execution.

How should this convention be documented?  Markdown file in the elvish repo?

**Comment by zzamboni** [comment]
It should be part of the `epm` documentation, if and when it's implemented.

@xiaq any thoughts?

**Comment by crinklywrappr** [comment]
I would like to provide an update.

I solved the problem which initially prompted me to create this ticket, and blogged about it [here](https://dev.to/crinklywrappr/generating-docs-from-tests-l64).

I also want to reiterate my support for a "blessed" Elvish test runner that's loosely integrated with the package manager. It's a standout component for a higher quality ecosystem than every other shell out there.

Same goes for docstrings and more reflection on builtins, but that's another discussion. ☺️
